### PR TITLE
feat: TreeMap migration

### PR DESCRIPTION
**Pull request by zhaoyongjie**
### SUMMARY

This PR intends to introduce an approach that migrates from v1 viz to v2 viz. The first PR implement TreeMap migration.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A

### TESTING INSTRUCTIONS

```shell
# upgrade
superset db upgrade

# downgrade
superset db downgrade
```

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/migrations/versions/2021-05-27_16-10_6f139c533bea_add_advanced_data_types_to_column_models.py [hunk 1]
```diff
@@ -14,10 +14,12 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
-"""adding_advanced_data_type.py
+"""adding advanced data type to column models
+
 Revision ID: 6f139c533bea
 Revises: cbe71abde154
 Create Date: 2021-05-27 16:10:59.567684
+
```
Comment (zhaoyongjie) [inline superset/migrations/versions/2021-05-27_16-10_6f139c533bea_add_advanced_data_types_to_column_models.py:22]: bycatch: pretty output

image

File: superset/utils/migrate_viz.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+import json
+from enum import Enum
+from typing import Any, Dict, Set, Type, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from superset.models.slice import Slice
+
+
+class MirateVizEnum(str, Enum):
```
[the remark this answers is not in the document]
Comment (zhaoyongjie) [reply superset/utils/migrate_viz.py:27]: Ah!! nice catch!

File: tests/unit_tests/conftest.py [hunk 3]
```diff
@@ -68,7 +69,9 @@ def app() -> Iterator[SupersetApp]:
     app = SupersetApp(__name__)
 
     app.config.from_object("superset.config")
-    app.config["SQLALCHEMY_DATABASE_URI"] = "sqlite://"
+    app.config["SQLALCHEMY_DATABASE_URI"] = (
+        os.environ.get("SUPERSET__SQLALCHEMY_DATABASE_URI") or "sqlite://"
+    )
```
Comment (zhaoyongjie) [inline tests/unit_tests/conftest.py:74]: Use integration test database instance for local debugging.

File: tests/unit_tests/utils/viz_migration/treemap_migration_test.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  "timeseries_limit_metric": "sum__num",
+  "treemap_ratio": 1.618033988749895,
+  "viz_type": "treemap"
+}
+"""
+
+treemap_processor = get_migrate_class[MigrateVizEnum.treemap]
+
+
+def test_treemap_migrate(app_context: SupersetApp) -> None:
```
Comment (michael-s-molina) [inline tests/unit_tests/utils/viz_migration/treemap_migration_test.py:63]: Maybe have separated tests for upgrade and downgrade?

File: superset/utils/migrate_viz.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            slc.viz_type = form_data_bak.get("viz_type")
+
+            query_context = json.loads(slc.query_context or "{}")
+            if "form_data" in query_context:
+                query_context["form_data"] = form_data_bak
+                slc.query_context = json.dumps(query_context)
+        return slc
+
+
+class MigrateTreeMap(MigrateViz):
```
Comment (michael-s-molina) [inline superset/utils/migrate_viz.py:109]: Maybe move `MigrateTreeMap` to its own file? I guess we'll have a bunch of these.

File: superset/utils/migrate_viz.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        if "viz_type" in self.data:
+            self.data["viz_type"] = self.target_viz_type
+
+        rv_data = {}
+        for (key, value) in self.data.items():
+            if key in self.mapping_keys and self.mapping_keys[key] in rv_data:
+                raise ValueError("Duplicate key in target viz")
+
+            if key in self.mapping_keys:
+                rv_data[self.mapping_keys[key]] = value
```
Comment (michael-s-molina) [inline superset/utils/migrate_viz.py:61]: I personally find this more readable, but it's not a blocker.

```suggestion
            if key in self.mapping_keys:
                mapped_key = self.mapping_keys[key] 
                if mapped_key in rv_data:
                   raise ValueError("Duplicate key in target viz")
                rv_data[mapped_key] = value          
```

File: tests/unit_tests/utils/viz_migration/treemap_migration_test.py (hunk 4 above)
Comment (zhaoyongjie) [reply tests/unit_tests/utils/viz_migration/treemap_migration_test.py:63]: the method of downgrade needs an updated slice so combine upgrade and downgrade in the same unit test case.

File: superset/utils/migrate_viz.py (hunk 5 above)
Comment (zhaoyongjie) [reply superset/utils/migrate_viz.py:109]: This is a one-time script. There is no need to split moudles.

Comment (michael-s-molina) [reply superset/utils/migrate_viz.py:109]: Won't we have other viz types to be migrated? If yes, are you going to add all the viz types to this file?

Comment (zhaoyongjie) [reply superset/utils/migrate_viz.py:109]: yep, all migrations are placed in this script.

Comment (zhaoyongjie) [reply superset/utils/migrate_viz.py:109]: Generally, _scripts_ are waterfalls, while _libary_ should be modular.

File: superset/utils/migrate_viz.py (hunk 6 above)
Comment (zhaoyongjie) [reply superset/utils/migrate_viz.py:61]: Exception handling is easier to maintain if it can be put together.

**Comment by michael-s-molina** [comment]
@zhaoyongjie @rusackas Shouldn't we disable TreeMap v1 after the migration? By that, I mean no more TreeMap v1 on Add Slice and Chart Type filter in the charts list. Right now, the user can still create the old version which kind of defeats the whole purpose of the migration.

**Comment by zhaoyongjie** [comment]
> @zhaoyongjie @rusackas Shouldn't we disable TreeMap v1 after the migration? By that, I mean no more TreeMap v1 on Add Slice and Chart Type filter in the charts list. Right now, the user can still create the old version which kind of defeats the whole purpose of the migration.

I intend to do a separated PR that removes treemap v1 after merged migration PR. the follow-up PR will 1) remove Legacy TreeMap from plugins, 2) remove legacy TreeMap from MainPresent.js.

File: superset/utils/migrate_viz.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    treemap = "treemap"
+
+
+class MigrateViz:
+    remove_keys: Set[str] = set()
+
+    mapping_keys: Dict[str, str] = {}
+
+    source_viz_type: str
+
```
Comment (john-bodley) [inline superset/utils/migrate_viz.py:39]:
```suggestion
```

File: superset/migrations/versions/2022-06-30_22-04_c747c78868b6_migrating_legacy_treemap.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    )
+    total = slices.count()
+    idx = 0
+    for slc in slices.yield_per(100):
+        try:
+            idx += 1
+            print(f"Upgrading ({idx}/{total}): {slc.slice_name}#{slc.id}")
+            new_viz = treemap_processor.upgrade(slc)
+            session.merge(new_viz)
+            session.commit()
```
[the remark this answers is not in the document]
Comment (zhaoyongjie) [reply superset/migrations/versions/2022-06-30_22-04_c747c78868b6_migrating_legacy_treemap.py]: done.

**Comment by zhaoyongjie** [comment]
@john-bodley thanks for the reviewing, I have addressed comments.

**Review by michael-s-molina — APPROVED** [review]

**Pull request merged**